GH-75586: Fix case where PATHEXT isn't applied to items in PATH (Windows) (#103179)
python/cpython: Doc/library/shutil.rst

   When no *path* is specified, the results of :func:`os.environ` are used,
   returning either the "PATH" value or a fallback of :attr:`os.defpath`.

   On Windows, the current directory is prepended to the *path* if
   the *mode* does not include ``os.X_OK``. When the *mode* does include
   ``os.X_OK``, the Windows API ``NeedCurrentDirectoryForExePathW`` will be
   consulted to determine if the current directory should be prepended to

>    *path*.

Reviewer: ```suggestion
   *path*. This will match the behavior of most shells on all platforms,
   preventing executables from being found in the current directory.
   For legacy shell behavior, pass *mode* without the ``os.X_OK`` flag.
```

Author: I don't think this is the case though. If the mode is executable, the curdir is added. Otherwise it asks `NeedCurrentDirectoryForExePath` if it should add curdir or not.

Windows behavior is still odd since it likely adds curdir for you.

Author: Tried to clarify a bit.

   Additionally, when finding the *cmd* in the *path*, the ``PATHEXT``
   environment variable is checked.  For example, if you call
   ``shutil.which("python")``, :func:`which` will search ``PATHEXT``
   to know that it should look for ``python.exe`` within the *path*

>    directories. For example, on Windows::

Reviewer: Maybe try rewording this to start with "Also on Windows, the ``PATHEXT`` variable is used to resolve commands that may not include an extension"? I agree, this is tough to word well, especially when we're trying to keep it relatively cross-platform.

      >>> shutil.which("python")
      'C:\\Python33\\python.EXE'

   Similar ``PATHEXT`` logic is also applied when a full path to a *cmd* is

>    given, though without an extension::

Reviewer: I think we should be able to merge this case into the earlier paragraph, especially since I'm _pretty sure_ this isn't correct? `PATHEXT` is always used, it's just that a full match without anything from `PATHEXT` will match first, so chances are if you provide the extension then it'll find that.

Author: Actually this wasn't the case prior to this changeset (it was the main reason I started looking at this issue).. See this example:

```
C:\Users\csm10495\Desktop>python
Python 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import shutil
>>> shutil.which("C:\\Python3\\python")
>>>
>>> exit()

C:\Users\csm10495\Desktop>cd cpython\

C:\Users\csm10495\Desktop\cpython>python
Running Release|x64 interpreter...
Python 3.12.0a6+ (heads/fix_which_pathext_windows-dirty:5fac84a, Apr  2 2023, 14:41:38) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import shutil
>>> shutil.which("C:\\Python3\\python")
'C:\\Python3\\python.EXE'
>>> exit()
```

So older Python didn't try PATHEXT with a fully given path on Windows.

      >> shutil.which("C:\\Python33\\python")
      'C:\\Python33\\python.EXE'

   .. versionadded:: 3.3

   .. versionchanged:: 3.8
      The :class:`bytes` type is now accepted.  If *cmd* type is
      :class:`bytes`, the result type is also :class:`bytes`.

   .. versionchanged:: 3.12

>       On Windows: ``NeedCurrentDirectoryForExePathW`` will be consulted
>       for non- ``os.X_OK`` modes to determine if the current working directory
>       should be prepended to *path*. Additionally, the ``PATHEXT`` environment
>       variable is now consulted when a full path to a cmd, minus the extension,
>       is given. Also, now a *cmd* with a matching ``PATHEXT`` extension will
>       be returned prior to one fully matching, if the fully matching one is
>       found later in *path*.

Reviewer: ```suggestion
      On Windows, queries allowing executables (``os.X_OK``) will now
      consult ``NeedCurrentDirectoryForExePathW`` to determine if the
      current working directory should be searched first. Additionally,
      the ``PATHEXT`` environment variable is now consulted even when
      *cmd* includes an extension. Finally, a *cmd* found with a
      ``PATHEXT`` extension in an earlier directory from *path* will
      now be returned ahead of any match found later in *path*.
```

Apologies for the whole rewrite, but with rewrapping text the suggestions started overlapping.

I _think_ I've captured it right, but double-check my working.

>    .. versionchanged:: 3.12
>       On Windows, the current directory is no longer prepended to the search
>       path if *mode* includes ``os.X_OK`` and WinAPI
>       ``NeedCurrentDirectoryForExePathW(cmd)`` is false; ``PATHEXT`` is used
>       now even when *cmd* includes a directory component or ends with an
>       extension that is in ``PATHEXT``; and filenames that have no extension
>       can be found now.

Reviewer 2: Mention the new CWD behavior:

```suggestion
   .. versionchanged:: 3.12
      On Windows, the current directory is no longer prepended to the search
      path if *mode* includes ``os.X_OK`` and WinAPI
      ``NeedCurrentDirectoryForExePathW(cmd)`` is false, else the current
      directory is prepended even if it is already in the search path;
      ``PATHEXT`` is used now even when *cmd* includes a directory component
      or ends with an extension that is in ``PATHEXT``; and filenames that
      have no extension can now be found.
```

   not include ``os.X_OK``. When the *mode* does include ``os.X_OK``, the
   Windows API ``NeedCurrentDirectoryForExePathW`` will be consulted to
   determine if the current directory should be prepended to *path*. To avoid
   consulting the current working directory for executables: set the environment
   variable ``NoDefaultCurrentDirectoryInExePath``.

   Also on Windows, the ``PATHEXT`` variable is used to resolve commands
   that may not already include an extension. For example, if you call
   ``shutil.which("python")``, :func:`which` will search ``PATHEXT``
   to know that it should look for ``python.exe`` within the *path*
   directories. For example, on Windows::

      >>> shutil.which("python")
      'C:\\Python33\\python.EXE'

   This is also applied when *cmd* is a path that contains a directory
   component::

      >> shutil.which("C:\\Python33\\python")
      'C:\\Python33\\python.EXE'

   .. versionadded:: 3.3

   .. versionchanged:: 3.8
      The :class:`bytes` type is now accepted.  If *cmd* type is
      :class:`bytes`, the result type is also :class:`bytes`.

   .. versionchanged:: 3.12

>    .. versionchanged:: 3.12

Reviewer: ```suggestion
   .. versionchanged:: 3.12
```

Hopefully this suggestion is against the current version.

>       On Windows, queries allowing executables (``os.X_OK``) will now
>       consult ``NeedCurrentDirectoryForExePathW`` to determine if the
>       current working directory should be searched first. Additionally,
>       the ``PATHEXT`` environment variable is now consulted even when
>       *cmd* includes a directory component. Finally, a *cmd* found with a
>       ``PATHEXT`` extension in an earlier directory from *path* will
>       now be returned ahead of any match found later in *path*.

Reviewer 2: ```suggestion
   .. versionchanged:: 3.12
      On Windows, the current directory is no longer prepended to the search
      path if *mode* includes ``os.X_OK`` and WinAPI
      ``NeedCurrentDirectoryForExePathW(cmd)`` is false; ``PATHEXT`` is used
      now even when *cmd* includes a directory component or ends with an
      extension that is in ``PATHEXT``; and filenames that have no extension
      can be found now.
```

   When no *path* is specified, the results of :func:`os.environ` are used,
   returning either the "PATH" value or a fallback of :attr:`os.defpath`.

   On Windows, the current directory is prepended to the *path* if

>    the *mode* does not include ``os.X_OK``. When the *mode* does include

Reviewer 2: I think "\*mode\*" is better than "the \*mode\*", given we're discussing a parameter name.

   ``os.X_OK``, the Windows API ``NeedCurrentDirectoryForExePathW`` will be
   consulted to determine if the current directory should be prepended to
   *path*. To avoid consulting the current working directory for executables,

>    and thereby, match the behavior of non-Windows shells: set the environment

Reviewer 2: I wouldn't mention "match the behavior of non-Windows shells". PowerShell doesn't search the current working directory, and CMD and `CreateProcessW()` don't either if `NoDefaultCurrentDirectoryInExePath` is defined.

   variable ``NoDefaultCurrentDirectoryInExePath``.

   Also on Windows, the ``PATHEXT`` variable is used to resolve commands

>    that may not already include an extension. For example, if you call

Reviewer 2: This example is described in detail and then shown explicitly in the REPL. I think it suffices just to show the example in the REPL.

Author: I've made it a bit less verbose.. maybe its ok now?

   ``shutil.which("python")``, :func:`which` will search ``PATHEXT``
   to know that it should look for ``python.exe`` within the *path*
   directories. For example, on Windows::

      >>> shutil.which("python")
      'C:\\Python33\\python.EXE'

>    Similar ``PATHEXT`` logic is also applied when a full path to a *cmd* is

Reviewer 2: "full path" or "fully qualified path" means an absolute path, but the logic here applies to any qualified path, or stated simply: "... also applied when \*cmd\* is a path that contains a directory component::"
